docs/api/configuration.rst
=================
Configuration API
=================

Overview
--------

99% of configuration use cases can be met with


.. currentmodule:: datacube.cfg

.. autosummary::

   :nosignatures:
   :toctree: generate/

   ODCConfig.get_environment

Full API
--------

.. currentmodule:: datacube.cfg

.. autoclass:: ODCConfig

   :members:

.. currentmodule:: datacube.cfg

.. autoclass:: ODCEnvironment

   :members:


docs/installation/database/configuration.rst

.. py:currentmodule:: datacube

ODC Configuration (Basics)
**************************

The Open Data Cube uses configuration files and/or environment variables to determine how to connect to databases.

Further functionality may be controlled by configuration in future releases.  (e.g. AWS/S3 access configuration,
rasterio configuration, etc.)

ODC Configuration Overview
==========================

When you first start a session with the Open Data Cube, you instantiate a
:py:class:`Datacube` object:

.. code-block:: python

   from datacube import Datacube

   dc = Datacube()

If you have access to many Open Data Cube databases, you may need to use several at once, e.g. to compare
the contents of dev and prod databases, or to combine data managed by different organisations.  In this
scenario, you instantiate a separate :py:class:`Datacube` object per environment:

.. code-block:: python

   from datacube import Datacube

   dc_prod = Datacube(env="prod")
   dc_dev  = Datacube(env="dev")

Environments can be read from a configuration file (e.g. an INI or YAML format file at :file:`~/.datacube.conf`) that
looks something like this:

.. code-block:: yaml

   # This is a YAML file and the # symbol marks comments
   default:
      # The 'default' environment is used if no environment is specified.
      # It is often convenient to define it as an alias to another environment
      alias: prod

   # You might have to copy configuration for system-wide environments from your system
   # configuration file.  (Probably at /etc/defaults/datacube.conf or /etc/datacube.conf)
   prod:
      # Production DB uses the legacy ODC index schema.
      index_driver: postgres
      # db_url is the easiest way to specify connection details
      db_url: postgresql://user:passwd@server.domain:5555/production_db
      db_connection_timeout: 30

   production:
      alias: prod

   dev:
      # Dev use the new PostGIS-based ODC index schema.
      index_driver: postgis
      db_url: postgresql://user:passwd@internal.server.domain:5555/development_db
      db_connection_timeout: 120

   development:
      alias: dev

   private:
      index_driver: postgis
      # Use OS ident authentication over a local named pipe.
      db_url: postgresql:///private

You can also inject new environments dynamically with environment variables, e.g.:

.. code-block:: python

   import os
   from datacube import Datacube
   os.environ["ODC_PRIVATE_INDEX_DRIVER"] = "postgis"
   os.environ["ODC_PRIVATE_DB_URL"] = "postgresql:///private"

   dc_private = Datacube(env="private")

Configuration Files
===================

Format
------

Configuration files may be provided in either INI or YAML format.  YAML is preferred
for consistency with ODC metadata files.  INI files can only support one level of nesting,
which is sufficient for current functionality - INI format may be deprecated for
configuration files in future releases if deeper nesting of configuration becomes
desirable for future functionality.

INI format configuration files are parsed with the Python standard library
configparser module.  Features supplied by that library are supported in ODC for
INI format configuration files only.  (e.g. a ``DEFAULT`` section whose
options are applied to all other sections unless over-ridden, and interpolation.)
Refer to the :py:mod:`configparser` documentation in the Python standard library
for more information.

Configuration Environments
--------------------------

A valid configuration file consists of one or more named environment definition sections.

Environment names must start with a lowercase letter and can only include lowercase
letters and digits.  (This restriction it to support generic environment variable
overrides, as discussed below.)

.. code-block:: ini
   :caption: Full INI Configuration Example

    ; Comments in INI files start with a semi-colon
    ; This config file defines two environments: 'main' and 'aux'.
    [main]
    index_driver: default
    db_database: datacube
    db_hostname: server.domain.com
    db_username: cube
    db_password: this_is_a_big_secret

    [aux]
    index_driver: default
    db_database: mydb
    ; Leaving the hostname blank uses a local socket.
    db_hostname:


.. code-block:: yaml
   :caption: Full YAML Configuration Example

    # Comments in YAML files start with a hash.
    # This config file defines two environments: 'main' and 'aux'.
    main:
      index_driver: default
      db_database: datacube
      db_hostname: server.domain.com
      db_username: cube
      db_password: this_is_a_big_secret

    aux:
      index_driver: default
      db_database: mydb
      # Leaving the hostname blank uses a local socket.
      db_hostname:

Configuration Options
---------------------

All supported configuration options are described here.  Configuration options are
specified per-environment.

.. confval:: alias

   **Cannot be used in conjunction with any other configuration option.**

   Normally an environment section in a configuration file defines a new
   environment.  If the ``alias`` configuration option is used, the section
   instead defines an alias for an existing environment.  If the alias option
   is present in a section, no other configuration options are permitted in
   that section.


   .. code-block::

      [default]
      ; The default environment is an alias for the "main" section.
      ; The 'main' environment can be accessed as either 'main' or 'default'.
      alias: main

      [main]
       index_driver: default
       db_database: datacube
       db_hostname: server.domain.com
       db_username: cube
       db_password: this_is_a_big_secret

.. confval:: index_driver

   Defines which index driver should be used to access the database index for
   this environment.

   The Open Data Cube currently supports 4 index drivers:

   - ``postgres`` Postgres index driver (aka ``default``, ``legacy``).  This
     is the old-style index driver, fully compatible with datacube-1.8.  This
     is the default value used if index_driver is not specified in the
     configuration.

     This index driver will not be available in datacube-2.0.

   - ``postgis`` Postgis index driver.  This is the new-style eo3-only index
     driver with support for spatial indexes.

   - ``memory`` In-memory index driver.  This index driver is currently
     compatible with the postgres driver, and stores all data temporarily in
     memory.  No persistent database is used.

   - ``null``  Null index driver.  If you are not using a database index at
     all, this might be an appropriate choice.

The ``null`` and ``memory`` index drivers take no further configuration. The
remaining configuration options only apply to the ``postgres`` and
``postgis`` index drivers:

.. confval:: db_connection_timeout

   **Only used for the 'postgres' and 'postgis' index drivers.**

   The database connection timeout, in seconds.

   Connections in the connection pool that are idle for more than the
   configured timeout are automatically closed.

   Defaults to 60.

.. confval:: db_url

   **Only used for the 'postgres' and 'postgis' index drivers.**

   Database connection details can be specified in a single option with the
   ``db_url`` field.  If a ``db_url`` is not provided, connection details can
   be specified with separate :confval:`db_hostname`, :confval:`db_port`, :confval:`db_database`,
   :confval:`db_username`, and :confval:`db_password` fields, as described below.

   If a `db_url` is provided, it takes precedence over the separate connection
   detail options.

   .. code-block:: ini
      :caption: INI Example showing :confval:`db_url`

      [default]
      index_driver: postgres
      ; Connect to database mydb on TCP port 5444 at server.domain, with username and password
      db_url: postgresql://username:password@server.domain:5444/mydb


   .. code-block:: yaml
      :caption: YAML Example showing :confval:`db_url`

      default:
        # Connect to database mydb over local socket with OS authentication.
        db_url: postgresql:///mydb

.. confval:: db_database

   **Only used for the 'postgres' and 'postgis' index drivers.**

   **Only used if :confval:`db_url` is not set.**

   The name of the database to connect to.  Defaults to ``"datacube"``.

.. confval:: db_hostname

   **Only used for the 'postgres' and 'postgis' index drivers.**

   **Only used if :confval:`db_url` is not set.**

   The hostname to connect to.  May be set to an empty string, in which case a
   local socket is used. Defaults to ``"localhost"`` if not set at all.

.. confval:: db_port

   **Only used for the 'postgres' and 'postgis' index drivers.**

   **Only used if :confval:`db_url` is not set.**

   The TCP port to connect to.  Defaults to 5432.  Not used when connecting over a local socket.

.. confval:: db_username

   **Only used for the 'postgres' and 'postgis' index drivers.**

   **Only used if :confval:`db_url` is not set.**

   The username to use when connecting to the database. Defaults to the
   username of the logged-in user on UNIX-like systems.

.. confval:: db_password

   .. admonition::
      Only used for the 'postgres' and 'postgis' index drivers.

      Only used if :confval:`db_url` is not set.

   The password to use when connecting to the database. Not used when
   connecting over a local socket.

.. confval:: db_iam_authentication

   **Only used for the 'postgres' and 'postgis' index drivers.**

   A boolean flag to indicate that IAM style authentication should be used
   instead of the supplied password.  (Recommended for cloud based database
   services like AWS RDS.)

   Defaults to False.

   .. code-block::
      :caption: Example showing :confval:`db_iam_authentication`

      [main]
      index_driver: postgis
      db_url: postgresql://user@server.domain:5432/main
      ; Use IAM authentication
      db_iam_authentication: yes

      [aux]
      index_driver: postgis
      db_url: postgresql:///aux
      db_iam_authentication: no

   YAML is a typed format and INI is not. Not all YAML boolean keywords will be
   recognised when they occur in INI files.  Using "yes" and "no" will work
   correctly for both formats.

   For IAM authentication to work, you must use the standard boto ``$AWS_*``
   environment variables to pass in your AWS identity and access key.

.. confval:: db_iam_timeout

   **Only used for the 'postgres' and 'postgis' index drivers.**

   **Only used when IAM authentication is activated.**

   How often (in seconds) a new IAM token should be generated.

   Defaults to 600 (10 minutes).

Need to know more?
==================

A full description of the ODC configuration engine can be found in :doc:`passing-configuration`.


datacube/cfg/cfg.py
# This file is part of the Open Data Cube, see https://opendatacube.org for more information
#
# Copyright (c) 2015-2024 ODC Contributors
# SPDX-License-Identifier: Apache-2.0
"""
Low level config path resolution, loading and multi-format parsing functions.

The default search path and default config also live here.
"""

import os
import warnings
from enum import Enum
from os import PathLike
from os.path import expanduser

from datacube.cfg.exceptions import ConfigException
from datacube.cfg.utils import ConfigDict, smells_like_ini, SemaphoreCallback

_DEFAULT_CONFIG_SEARCH_PATH = [
    "datacube.conf",      # i.e. in the current working directory.
    expanduser("~/.datacube.conf"),   # i.e. in user's home directory.
    # Check if we are running under a Windows and use Windowsy default paths?
    "/etc/default/datacube.conf",  # Preferred location for global config file
    "/etc/datacube.conf",          # Legacy location for global config file
]

_DEFAULT_CONF = """
default:
   db_hostname: ''
   db_database: datacube
   index_driver: default
   db_connection_timeout: 60
"""


def find_config(paths_in: None | str | PathLike | list[str | PathLike],
                default_cb: SemaphoreCallback | None = None) -> str:
    """
    Given a file system path, or a list of file system paths, return the contents of the first file
    in the list that can be read as a string.

    If "None" is passed in the default config search path is used:

        "datacube.conf"               (i.e. 'datacube.conf' in the current working directory.)
        "~/.datacube.conf"            (i.e. '.datacube.conf' in the user's home directory.)
        "/etc/default/datacube.conf"  (Preferred location for global config file)
        "/etc/datacube.conf"          (Legacy location for global config file)

    If a path or list of paths was passed in, AND no readable file could be found, a ConfigException is raised.
    If None was passed in, AND no readable file could be found, a default configuration text is returned.

    :param paths_in: A file system path, or a list of file system paths, or None.
    :return: The contents of the first readable file found.
    """
    using_default_paths: bool = False
    paths: list[str | PathLike] = []
    if paths_in is None:
        if os.environ.get("ODC_CONFIG_PATH"):
            paths.extend(os.environ["ODC_CONFIG_PATH"].split(':'))
        elif os.environ.get("DATACUBE_CONFIG_PATH"):
            warnings.warn(
                "Datacube config path being determined by legacy $DATACUBE_CONFIG_PATH environment variable. "
                "This environment variable is deprecated and the behaviour of it has changed somewhat since datacube "
                "1.8.x.   Please refer to the documentation for details and switch to $ODC_CONFIG_PATH"
            )
            paths.extend(os.environ["DATACUBE_CONFIG_PATH"].split(':'))
        else:
            paths.extend(_DEFAULT_CONFIG_SEARCH_PATH)
            using_default_paths = True
    elif isinstance(paths_in, str) or isinstance(paths_in, PathLike):
        paths.append(paths_in)
    else:
        paths.extend(paths_in)

    for path in paths:
        try:
            with open(path, "r") as fp:
                return fp.read()
        except PermissionError:
            continue
        except FileNotFoundError:
            continue

    if using_default_paths:
        if default_cb is not None:
            default_cb()
        return _DEFAULT_CONF

    raise ConfigException("No configuration file found in the provided locations")


class CfgFormat(Enum):
    """
    An Enum class for config file formats.
    """
    AUTO = 0   # Format unspecified - autodetect
    INI = 1
    YAML = 2
    JSON = 2   # JSON is a subset of YAML


def parse_text(cfg_text: str, fmt: CfgFormat = CfgFormat.AUTO) -> ConfigDict:
    """
    Parse a string of text in INI, JSON or YAML format into a raw dictionary.

    Raises a ConfigException if the file cannot be parsed.

    :param cfg_text: Configuration string in INI, YAML or JSON format
    :param fmt: Whether to use the ini or yaml/json parser. By default autodetects file format.
    :return: A raw config dictionary
    """
    raw_config = {}
    if fmt == fmt.INI or (
            fmt == fmt.AUTO and smells_like_ini(cfg_text)):
        # INI parsing
        import configparser
        try:
            ini_config = configparser.ConfigParser()
            ini_config.read_string(cfg_text)
            for section in ini_config.sections():
                sect = {}
                for key, value in ini_config.items(section):
                    sect[key] = value

                raw_config[section] = sect
        except configparser.Error as e:
            raise ConfigException(f"Invalid INI file: {e}")
    else:
        # YAML/JSON parsing
        import yaml
        try:
            raw_config = yaml.load(cfg_text, Loader=yaml.Loader)
        except yaml.parser.ParserError as e:
            raise ConfigException(f"Invalid YAML file:{e}")

    return raw_config
